src/db/car/forest.rs

```diff
@@ -59,76 +59,65 @@ use cid::Cid;
 use futures::{Stream, TryStream, TryStreamExt as _};
 use fvm_ipld_blockstore::Blockstore;
 use fvm_ipld_encoding::to_vec;
-use parking_lot::Mutex;
+use parking_lot::RwLock;
+use positioned_io::{Cursor, ReadAt};
+
 use std::sync::Arc;
 use std::task::Poll;
 use std::{
     io,
-    io::{Read, Seek, SeekFrom, Write},
+    io::{Read, Write},
 };
 use tokio::io::{AsyncWrite, AsyncWriteExt};
 use tokio_util::codec::{Decoder, Encoder as _};
 
-pub trait ReaderGen<V>: Fn() -> io::Result<V> + Send + Sync + 'static {}
-impl<ReaderT, X: Fn() -> io::Result<ReaderT> + Send + Sync + 'static> ReaderGen<ReaderT> for X {}
+pub trait ReaderGen<V>: Fn() -> io::Result<(V, u64)> + Send + Sync + 'static {}
+impl<ReaderT, X: Fn() -> io::Result<(ReaderT, u64)> + Send + Sync + 'static> ReaderGen<ReaderT>
+    for X
+{
+}
 
 pub struct ForestCar<ReaderT> {
     // Multiple `ForestCar` structures may share the same cache. The cache key is used to identify
     // the origin of a cached z-frame.
     cache_key: CacheKey,
-    indexed: Mutex<io::Result<CarIndex<ReaderT>>>,
-    new_reader: Arc<dyn ReaderGen<CarIndex<ReaderT>>>,
-    frame_cache: Arc<Mutex<ZstdFrameCache>>,
-    write_cache: Arc<Mutex<ahash::HashMap<Cid, Vec<u8>>>>,
+    indexed: CarIndex<ReaderT>,
+    frame_cache: Arc<RwLock<ZstdFrameCache>>,
+    write_cache: Arc<RwLock<ahash::HashMap<Cid, Vec<u8>>>>,
     roots: Vec<Cid>,
 }
 
-// Cloning a `ForestCar` duplicates the reader. If the reader is a [`std::fs::File`] then a new
-// handle will be created pointing to the same underlying file. Each `ForestCar` clone can be
-// accessed in parallel and will share the same z-frame cache.
-impl<ReaderT: Read + Seek> Clone for ForestCar<ReaderT> {
-    fn clone(&self) -> Self {
-        let index: io::Result<CarIndex<ReaderT>> = (self.new_reader)();
-        ForestCar {
-            cache_key: self.cache_key,
-            indexed: Mutex::new(index),
-            new_reader: Arc::clone(&self.new_reader),
-            frame_cache: Arc::clone(&self.frame_cache),
-            write_cache: Arc::clone(&self.write_cache),
-            roots: self.roots.clone(),
-        }
-    }
-}
-
 impl<ReaderT: super::CarReader> ForestCar<ReaderT> {
     pub fn new(mk_reader: impl ReaderGen<ReaderT>) -> io::Result<Self> {
-        let mut reader = mk_reader()?;
+        let (reader, file_size) = mk_reader()?;
 
-        reader.seek(SeekFrom::End(-(ForestCarFooter::SIZE as i64)))?;
         let mut footer_buffer = [0; ForestCarFooter::SIZE];
-        reader.read_exact(&mut footer_buffer)?;
+
+        // Make sure we don't panic when the filesize is less than the footer size.
+        let read_offset = match file_size.checked_sub(ForestCarFooter::SIZE as u64) {
```

Comment:
```suggestion
        let read_offset = file_size.saturating_sub(ForestCarFooter::SIZE as u64);
```

Reply:
This is outdated, not going to be needed anymore.